src/default.jl

```diff
@@ -1,209 +1,136 @@
-## Default algorithm
-
-# Allows A === nothing as a stand-in for dense matrix
-function defaultalg(A, b)
-    if A isa DiffEqArrayOperator
-        A = A.A
-    end
-
-    # Special case on Arrays: avoid BLAS for RecursiveFactorization.jl when
-    # it makes sense according to the benchmarks, which is dependent on
-    # whether MKL or OpenBLAS is being used
-    if (A === nothing && !(b isa GPUArraysCore.AbstractGPUArray)) || A isa Matrix
-        if (A === nothing || eltype(A) <: Union{Float32, Float64, ComplexF32, ComplexF64}) &&
-           ArrayInterfaceCore.can_setindex(b)
-            if length(b) <= 10
-                alg = GenericLUFactorization()
-            elseif (length(b) <= 100 || (isopenblas() && length(b) <= 500)) &&
-                   eltype(A) <: Union{Float32, Float64}
-                alg = RFLUFactorization()
-                #elseif A === nothing || A isa Matrix
-                #    alg = FastLUFactorization()
-            else
-                alg = LUFactorization()
-            end
-        else
-            alg = LUFactorization()
-        end
+function defaultalg(A::DiffEqArrayOperator, b, assumptions::OperatorAssumptions)
+    defaultalg(A.A, b, assumptions)
+end
 
-        # These few cases ensure the choice is optimal without the
-        # dynamic dispatching of factorize
-    elseif A isa Tridiagonal
-        alg = GenericFactorization(; fact_alg = lu!)
-    elseif A isa SymTridiagonal
-        alg = GenericFactorization(; fact_alg = ldlt!)
-    elseif A isa SparseMatrixCSC
-        if length(b) <= 10_000
-            alg = KLUFactorization()
-        else
-            alg = UMFPACKFactorization()
-        end
+# Ambiguity handling
+function defaultalg(A::DiffEqArrayOperator, b, assumptions::OperatorAssumptions{nothing})
+    defaultalg(A.A, b, assumptions)
+end
 
-        # This catches the cases where a factorization overload could exist
-        # For example, BlockBandedMatrix
-    elseif A !== nothing && ArrayInterfaceCore.isstructured(A)
-        alg = GenericFactorization()
+function defaultalg(A, b, ::OperatorAssumptions{nothing})
+    issquare = size(A, 1) == size(A, 2)
+    defaultalg(A, b, OperatorAssumptions(Val(issquare)))
```

Comment:
This seems type unstable. Is that intentional, or does Julia always union split here?

Reply:
We need to have some point where a type choice is made, because then it sets the algorithm type. Things like the ODE solver would just set the assumptions at the type level.